presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java

```diff
@@ -123,66 +120,63 @@ public Block readBlock(Type type)
             }
         }
 
-        assureVectorSize();
-
         BlockBuilder builder = type.createBlockBuilder(null, nextBatchSize);
-        while (nextBatchSize > 0) {
-            int subBatchSize = min(nextBatchSize, MAX_BATCH_SIZE);
-            if (presentStream == null) {
-                if (dataStream == null) {
-                    throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Value is not null but data stream is not present");
+
+        if (presentStream == null) {
+            if (dataStream == null) {
+                throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Value is not null but data stream is not present");
+            }
+            if (inDictionaryStream == null) {
+                //dataStream.nextIntVector(nextBatchSize, idsVector, 0);
+                for (int i = 0; i < nextBatchSize; i++) {
+                    type.writeLong(builder, dictionary[((int) dataStream.next())]);
                 }
-                Arrays.fill(nullVector, false);
-                dataStream.nextLongVector(subBatchSize, dataVector);
             }
             else {
-                int nullValues = presentStream.getUnsetBits(subBatchSize, nullVector);
-                if (nullValues != subBatchSize) {
-                    if (dataStream == null) {
-                        throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Value is not null but data stream is not present");
+                for (int i = 0; i < nextBatchSize; i++) {
+                    long id = dataStream.next();
+                    if (inDictionaryStream.nextBit()) {
+                        type.writeLong(builder, dictionary[(int) id]);
+                    }
+                    else {
+                        type.writeLong(builder, id);
                     }
-                    dataStream.nextLongVector(subBatchSize, dataVector, nullVector);
                 }
             }
-
-            if (inDictionaryStream == null) {
-                Arrays.fill(inDictionaryVector, true);
+        }
+        else {
+            // presentStream != null
```

Comment:
Instead of a physical comment a logical `Data has nulls` might be easier for the reader